pkg/controllers/provisioning/scheduling/topologydomaingroup.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+// TopologyDomainGroup tracks the domains for a single topology. Additionally, it tracks the taints associated with
+// each of these domains. This enables us to determine which domains should be considered by a pod if its
+// NodeTaintPolicy is honor.
+type TopologyDomainGroup map[string][]sets.Set[v1.Taint]
+
+func NewTopologyDomainGroup() TopologyDomainGroup {
+	return map[string][]sets.Set[v1.Taint]{}
+}
+
+// Insert either adds a new domain to the TopologyDomainGroup or updates an existing domain.
+func (t TopologyDomainGroup) Insert(domain string, taints ...v1.Taint) {
+	// Note: This could potentially be improved by removing any set of which the new set of taints is a proper subset.
+	// Currently this is only handled when the incoming set is the empty set.
+	if _, ok := t[domain]; !ok || len(taints) == 0 {
+		t[domain] = []sets.Set[v1.Taint]{sets.New(taints...)}
+		return
+	}
+	if t[domain][0].Len() == 0 {
+		// The domain already contains a set of taints which is the empty set, therefore this domain can be considered
+		// by all pods, regardless of their tolerations. There is no longer a need to track new sets of taints.
+		return
+	}
+	t[domain] = append(t[domain], sets.New(taints...))
+}
+
+// GetDomains returns a slice of all the domains in the topology tracked by this TopologyDomainGroup.
+func (t TopologyDomainGroup) GetDomains() []string {
+	return lo.Keys(t)
+}
+
+// GetToleratedDomains returns a slice of domains that the pod can potentially schedule to, given its tolerations.
+func (t TopologyDomainGroup) GetToleratedDomains(pod *v1.Pod) []string {
+	return lo.Filter(lo.Keys(t), func(domain string, _ int) bool {
+		for _, taints := range t[domain] {
+			if err := scheduling.Taints(taints.UnsortedList()).ToleratesPod(pod); err == nil {
```

Comment:
If we can void calling keys/unsortedlisted & friends in the scheduling code, it's better.  They construct new lists and we spend extra time on GC that we don't need to.